send-transaction-service/src/send_transaction_service.rs

```diff
@@ -109,100 +147,123 @@ impl SendTransactionService {
+        exit: Arc<AtomicBool>,
     ) -> JoinHandle<()> {
-        let mut last_status_check = Instant::now();
+        let mut last_batch_sent = Instant::now();
         let mut last_leader_refresh = Instant::now();
         let mut transactions = HashMap::new();
 
+        info!(
+            "Starting send-transaction-service::receive_txn_thread with config {:?}",
+            config
+        );
         if let Some(leader_info) = leader_info.as_mut() {
             leader_info.refresh_recent_peers();
         }
         connection_cache::set_use_quic(config.use_quic);
         Builder::new()
-            .name("send-tx-sv2".to_string())
+            .name("send-tx-receive".to_string())
             .spawn(move || loop {
-                match receiver.recv_timeout(Duration::from_millis(1000.min(config.retry_rate_ms))) {
-                    Err(RecvTimeoutError::Disconnected) => break,
+                let recv_timeout_ms = config.batch_send_rate_ms;
+                match receiver.recv_timeout(Duration::from_millis(1000.min(recv_timeout_ms))) {
+                    Err(RecvTimeoutError::Disconnected) => {
+                        info!("Terminating send-transaction-service.");
+                        exit.store(true, Ordering::Relaxed);
+                        break;
+                    }
                     Err(RecvTimeoutError::Timeout) => {}
                     Ok(transaction_info) => {
                         inc_new_counter_info!("send_transaction_service-recv-tx", 1);
-                        let transactions_len = transactions.len();
                         let entry = transactions.entry(transaction_info.signature);
+                        let mut new_txn = false;
                         if let Entry::Vacant(_) = entry {
-                            let addresses = leader_info.as_ref().map(|leader_info| {
-                                leader_info.get_leader_tpus(config.leader_forward_count)
-                            });
-                            let addresses = addresses
-                                .map(|address_list| {
-                                    if address_list.is_empty() {
-                                        vec![&tpu_address]
-                                    } else {
-                                        address_list
-                                    }
-                                })
-                                .unwrap_or_else(|| vec![&tpu_address]);
-                            for address in addresses {
-                                Self::send_transaction(address, &transaction_info.wire_transaction);
-                            }
-                            if transactions_len < MAX_TRANSACTION_QUEUE_SIZE {
-                                inc_new_counter_info!("send_transaction_service-insert-tx", 1);
+                            if !retry_transactions
+                                .lock()
```

Comment:
Yes. I thought about it. I do not think it will make a difference with current implementation. Read and Write will block each other even in the RwLock case. RwLock is only useful in the case of multiple reader case which is not true currently. We have a single receiver thread which is contending with the retry thread.

Reply:
yeah I was thinking that taking the read side of rwlock was cheaper than mutex even in the single reader case. maybe inaccurate